songpal/method.py

```diff
@@ -152,18 +152,78 @@ async def __call__(self, *args, **kwargs):
 
     @property
     def inputs(self) -> Dict[str, type]:
-        """Input parameters for this method."""
-        return self.signature.input
+        """Input parameters for this method version."""
+        return (
+            self.signatures[self._version].input
+            if self.signatures[self._version].input is not None
+            else ""
+        )
+
+    @property
+    def latest_supported_version(self) -> str:
+        """Latest version supported by this method."""
+        return (
+            sorted(self._supported_versions, reverse=True)[0]
+            if len(self._supported_versions) > 0
+            else None
+        )
 
     @property
     def outputs(self) -> Dict[str, type]:
-        """Output parameters for this method."""
-        return self.signature.output
+        """Output parameters for this method version."""
+        return (
+            self.signatures[self._version].output
+            if self.signatures[self._version].output is not None
+            else ""
+        )
+
+    @property
+    def signature(self) -> MethodSignature:
+        """Method version signature."""
+        return self.signatures[self._version]
+
+    @property
+    def supported_versions(self) -> Set[str]:
+        """List of supported version numbers for this method."""
+        return self._supported_versions
+
+    @property
+    def version(self) -> str:
+        """Method signature version number."""
+        return self._version
+
+    def add_supported_version(self, version: str):
+        """Add a supported version number for this method."""
+        self._supported_versions.add(version)
+
+    def supports_version(self, version: str) -> bool:
+        """Is this method version supported."""
+        return True if (version in self._supported_versions) else False
```

Comment:
```suggestion
        return version in self._supported_versions
```

Reply:
and then do

```
if method.use_version("1.2"):
    res = await method({})
```

instead of
```
if method.supports_version("1.2"):
    method.use_version("1.2")
    res = await method({})
```

Reply:
Just popped in my mind, that making the `__call__` to accept a `version` kwarg would potentially be even cleaner and make the method stateless? That'd require passing the wanted version always to every call, and might conflict with params?

Reply:
It would be, and I originally went down a similar path, but I didn't know how to handle param differences.

```
if method.use_version("1.2"):
    res = await method({})
else:
    res = await method()
```

Reply:
Ah, indeed. I have a dislike on side-effects as they may be confusing, but it's what it is :-)